<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Residual-based SHAP Analysis and Visualization in Spatial Transcriptomics — SpaPheno_SHAP_residual_analysis • SpaPheno</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Residual-based SHAP Analysis and Visualization in Spatial Transcriptomics — SpaPheno_SHAP_residual_analysis"><meta name="description" content="Performs residual-based SHAP analysis for a selected feature, identifying
outliers based on SHAP residuals and visualizing them in both spatial and dependence plots."><meta property="og:description" content="Performs residual-based SHAP analysis for a selected feature, identifying
outliers based on SHAP residuals and visualizing them in both spatial and dependence plots."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaPheno</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/SpaPheno-tutorial.html">SpaPheno Tutorial: Spatially-Informed Phenotype Prediction and Interpretation Using Single-Cell Data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/DuanLab1/SpaPheno/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Residual-based SHAP Analysis and Visualization in Spatial Transcriptomics</h1>
      <small class="dont-index">Source: <a href="https://github.com/DuanLab1/SpaPheno/blob/HEAD/R/SpaPheno_SHAP_residual_analysis.R"><code>R/SpaPheno_SHAP_residual_analysis.R</code></a></small>
      <div class="d-none name"><code>SpaPheno_SHAP_residual_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs residual-based SHAP analysis for a selected feature, identifying
outliers based on SHAP residuals and visualizing them in both spatial and dependence plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SpaPheno_SHAP_residual_analysis</span><span class="op">(</span></span>
<span>  <span class="va">shap_df</span>,</span>
<span>  <span class="va">feature_name</span>,</span>
<span>  <span class="va">coordinate_df</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"SHAP Residual Distribution"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-shap-df">shap_df<a class="anchor" aria-label="anchor" href="#arg-shap-df"></a></dt>
<dd><p>A data.frame containing SHAP results with at least the columns:
<code>feature</code>, <code>phi</code>, <code>feature.value</code>, and <code>sample</code>.</p></dd>


<dt id="arg-feature-name">feature_name<a class="anchor" aria-label="anchor" href="#arg-feature-name"></a></dt>
<dd><p>Character; the feature (e.g., gene) to analyze.</p></dd>


<dt id="arg-coordinate-df">coordinate_df<a class="anchor" aria-label="anchor" href="#arg-coordinate-df"></a></dt>
<dd><p>A data.frame of spatial coordinates with rownames matching <code>sample</code> IDs;
must contain columns named <code>X</code> and <code>Y</code>.</p></dd>


<dt id="arg-size">size<a class="anchor" aria-label="anchor" href="#arg-size"></a></dt>
<dd><p>Numeric; point size for plotting. Default is 1.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>Character; title for the spatial residual plot. Default is <code>"SHAP Residual Distribution"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p><ul><li><p><code>spatial_plot</code>: A <code>ggplot2</code> object showing spatial distribution of SHAP residuals.</p></li>
<li><p><code>dependence_plot</code>: A SHAP dependence plot with residual outlier groups.</p></li>
<li><p><code>residual_table</code>: A data.frame with residuals, Z-scores, and group annotations.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function regresses SHAP values on feature values to obtain residuals, then uses
Z-score normalization to identify spatial units with unusually high or low SHAP effects
that are not explained by the raw expression. The results are visualized both spatially
and as enhanced SHAP dependence plots.</p>
<p>Residual outliers are defined as:</p><ul><li><p>Z ≥ 2: High residual</p></li>
<li><p>Z ≤ -2: Low residual</p></li>
<li><p>Otherwise: Normal</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Bin Duan</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"osmFISH_metadata_cellType"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"osmFISH_bulk_decon"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"osmFISH_bulk_pheno"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"osmFISH_bulk_coordinate"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">PhenoResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="SpatialPhenoMap.html">SpatialPhenoMap</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  bulk_decon <span class="op">=</span> <span class="va">osmFISH_bulk_decon</span>,</span></span>
<span class="r-in"><span>  bulk_pheno <span class="op">=</span> <span class="va">osmFISH_bulk_pheno</span>,</span></span>
<span class="r-in"><span>  family <span class="op">=</span> <span class="st">"binomial"</span>,</span></span>
<span class="r-in"><span>  coord <span class="op">=</span> <span class="va">osmFISH_bulk_coordinate</span>,</span></span>
<span class="r-in"><span>  resolution <span class="op">=</span> <span class="st">"single_cell"</span>,</span></span>
<span class="r-in"><span>  sample_information_cellType <span class="op">=</span> <span class="va">osmFISH_metadata_cellType</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pred_result</span> <span class="op">&lt;-</span> <span class="va">PhenoResult</span><span class="op">$</span><span class="va">pred_score</span></span></span>
<span class="r-in"><span><span class="va">phenoPlus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">pred_result</span><span class="op">[</span><span class="va">pred_result</span><span class="op">$</span><span class="va">label</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"phenotype+"</span>, <span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="va">PhenoResult</span><span class="op">$</span><span class="va">model</span></span></span>
<span class="r-in"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">PhenoResult</span><span class="op">$</span><span class="va">cell_type_distribution</span><span class="op">[</span><span class="va">phenoPlus</span>, <span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># This step took a very long time</span></span></span>
<span class="r-in"><span><span class="va">shap_test_plus</span> <span class="op">&lt;-</span> <span class="fu"><a href="compute_shap_spatial.html">compute_shap_spatial</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  model <span class="op">=</span> <span class="va">model</span>,</span></span>
<span class="r-in"><span>  X_bulk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">osmFISH_bulk_decon</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  y_bulk <span class="op">=</span> <span class="va">osmFISH_bulk_pheno</span>,</span></span>
<span class="r-in"><span>  X_spatial <span class="op">=</span> <span class="va">X</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">resi_result</span> <span class="op">&lt;-</span> <span class="fu">SpaPheno_SHAP_residual_analysis</span><span class="op">(</span></span></span>
<span class="r-in"><span>shap_df <span class="op">=</span> <span class="va">shap_test_plus</span>,</span></span>
<span class="r-in"><span>feature_name <span class="op">=</span> <span class="st">"Perivascular.Macrophages"</span>,</span></span>
<span class="r-in"><span>coordinate_df <span class="op">=</span> <span class="va">test_coordinate</span>, size <span class="op">=</span> <span class="fl">0.8</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">resi_hot</span> <span class="op">&lt;-</span> <span class="va">resi_result</span><span class="op">$</span><span class="va">residual_table</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">resi_hot</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">resi_hot</span><span class="op">$</span><span class="va">phi_resid_z</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span>, <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="SpaPheno_SHAP_waterfall_plot.html">SpaPheno_SHAP_waterfall_plot</a></span><span class="op">(</span><span class="va">shap_test_plus</span>, <span class="st">"cell_5593"</span>, top_n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">resi_result</span><span class="op">$</span><span class="va">dependence_plot</span></span></span>
<span class="r-in"><span><span class="va">resi_result</span><span class="op">$</span><span class="va">spatial_plot</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Duan, Hua Zou.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

